common/persistence/dlq_metrics_emitter.go

```diff
@@ -0,0 +1,166 @@
+// if the history service it currently run hosts shard 1.
+type (
+	DLQMetricsEmitter struct {
+		status                  int32
+		shutdownCh              chan struct{}
+		metricsHandler          metrics.Handler
+		logger                  log.Logger
+		emitMetricsTimer        *time.Ticker
+		historyTaskQueueManager HistoryTaskQueueManager
+		historyServiceResolver  membership.ServiceResolver
+		hostInfoProvider        membership.HostInfoProvider
+		taskCategoryRegistry    tasks.TaskCategoryRegistry
+	}
+)
+
+func NewDLQMetricsEmitter(
+	metricsHandler metrics.Handler,
+	logger log.Logger,
+	manager HistoryTaskQueueManager,
+	historyServiceResolver membership.ServiceResolver,
+	hostInfoProvider membership.HostInfoProvider,
+	taskCategoryRegistry tasks.TaskCategoryRegistry,
+) *DLQMetricsEmitter {
+	return &DLQMetricsEmitter{
+		status:                  common.DaemonStatusInitialized,
+		shutdownCh:              make(chan struct{}),
+		metricsHandler:          metricsHandler,
+		emitMetricsTimer:        time.NewTicker(emitDLQMetricsInterval),
+		logger:                  logger,
+		historyTaskQueueManager: manager,
+		historyServiceResolver:  historyServiceResolver,
+		hostInfoProvider:        hostInfoProvider,
+		taskCategoryRegistry:    taskCategoryRegistry,
+	}
+}
+
+func (s *DLQMetricsEmitter) Start() {
+	if !atomic.CompareAndSwapInt32(&s.status, common.DaemonStatusInitialized, common.DaemonStatusStarted) {
+		return
+	}
+	go s.emitMetricsLoop()
+}
+
+func (s *DLQMetricsEmitter) Stop() {
+	if !atomic.CompareAndSwapInt32(&s.status, common.DaemonStatusStarted, common.DaemonStatusStopped) {
+		return
+	}
+	close(s.shutdownCh)
+	s.emitMetricsTimer.Stop()
+}
+
+func (s *DLQMetricsEmitter) emitMetricsLoop() {
+	for {
+		select {
+		case <-s.shutdownCh:
+			return
+		case <-s.emitMetricsTimer.C:
+			if !s.shouldEmitMetrics() {
+				continue
+			}
+			categories := s.taskCategoryRegistry.GetCategories()
+			messageCounts := make(map[int]int64)
+			for category := range categories {
+				messageCounts[category] = 0
+			}
+			queues, err := s.getDLQList()
+			if err != nil {
+				s.logger.Error("Failed to list DLQs to emit metrics", tag.Error(err))
+				continue
+			}
+			for _, q := range queues {
+				category, err := GetHistoryTaskQueueCategoryID(q.QueueName)
+				if err != nil {
+					s.logger.Error("Failed to process DLQ queue name", tag.Error(err))
+				}
+				messageCounts[category] += q.MessageCount
+			}
+			for id, count := range messageCounts {
+				category, ok := categories[id]
+				if !ok {
+					s.logger.Error("Failed to find category from ID ", tag.TaskCategoryID(id))
+				}
+				metrics.DLQMessageCount.With(s.metricsHandler).Record(float64(count), metrics.TaskCategoryTag(category.Name()))
+			}
+		}
+	}
+}
+
+func (s *DLQMetricsEmitter) getDLQList() ([]QueueInfo, error) {
+	var queues []QueueInfo
+	var nextPageToken []byte
+	for {
+		resp, err := s.historyTaskQueueManager.ListQueues(context.Background(), &ListQueuesRequest{
```

Comment:
we should mark the request as low priority. `SystemPreemptableCallerInfo`

Reply:
Done.